daemon/src/engine/server/CryptoChallenge.h

```diff
@@ -0,0 +1,203 @@
+     */
+    bool ValidAt( const TimePoint& time ) const
+    {
+        return created + timeout >= time;
+    }
+
+    /*
+     * Whether two challenges match
+     */
+    bool Matches( const Challenge& other ) const
+    {
+        if ( challenge != other.challenge )
+            return false;
+
+        return NET_CompareAdr( source, other.source );
+    }
+
+    /*
+     * Challenge as a hex string
+     */
+    std::string String() const
+    {
+        return Crypto::String( Crypto::Encoding::HexEncode( challenge ) );
+    }
+
+    /*
+     * Time when this challenge has been created
+     */
+    TimePoint Created() const
+    {
+        return created;
+    }
+
+    /*
+     * Duration this challenge has been alive for
+     */
+    Duration Lifetime() const
+    {
+        return Clock::now() - created;
+    }
+
+    /*
+     * Checks whether a string is a proper challenge representation
+     * (ie: something that String() might have returned
+     */
+    static bool ValidString(const std::string& challenge);
+
+private:
+    /*
+     * Generates a random challenge
+     */
+    std::string GenerateString();
+
+    TimePoint    created;
+    Duration     timeout;
+    Crypto::Data challenge;
+    netadr_t     source;
+
+};
+
+class ChallengeManager
+{
+public:
+    static ChallengeManager& Get()
+    {
+        static ChallengeManager singleton;
+        return singleton;
+    }
+
+    std::size_t MaxChallenges() const;
+
+    /*
+     * Generates a challenge for the given address and returns the challenge string
+     */
+    std::string GenerateChallenge( const netadr_t& source );
+
+    /*
+     * Add a challenge to the pool
+     */
+    void Push( const Challenge& challenge );
+
+    /*
+     * Check if a challenge matches any of the registered ones
+     * If it does and it's valid, it'll return true and remove it
+     *
+     * If ping is not null, it will be set to the challenge ping time
+     */
+    bool Match( const Challenge& challenge, Challenge::Duration* ping = nullptr );
+
+
+    /*
+     * Removes all active challenges
+     */
+    void Clear();
+
+    /*
+     * Matches a challenge, the challenge string should be equal to what
+     * GenerateChallenge() returned for the same source.
+     *
+     * If ping is not null, it will be set to the challenge ping time
+     */
+    bool MatchString( const netadr_t& source,
+                      const std::string& challenge,
+                      Challenge::Duration* ping = nullptr);
+
+private:
+    ChallengeManager() = default;
+    ChallengeManager( const ChallengeManager& ) = delete;
+    ChallengeManager& operator=( const ChallengeManager& ) = delete;
+
+    std::unique_lock<std::mutex> Lock()
+    {
+        return std::unique_lock<std::mutex>{mutex};
+    }
+
+    /*
+     * Removes outdated challenges
+     * PRE: The caller has acquired a lock on mutex
+     */
+    void Cleanup();
+
+    std::mutex mutex;
```

Comment:
Likewise, this is going to be used from the main thread only so no need for locking.